hw/top_earlgrey/util/opentitan_earlgrey_flash_size_reduce.py

```diff
@@ -23,46 +24,25 @@ def main():
 
     # Get path to top-level directory
     top_path = os.path.normpath(os.path.join(os.path.dirname(__file__), "../../.."))
+    top_hjson = top_path + "/hw/top_earlgrey/data/top_earlgrey.hjson"
 
-    # Change the following expressions in the following source files.
-    files = [
-        top_path + "/sw/device/lib/flash_ctrl.h",
-        top_path + "/hw/top_earlgrey/rtl/top_pkg.sv",
-        top_path + "/hw/top_earlgrey/data/top_earlgrey.hjson"
-    ]
-    match = [
-        r"(#define\s+FLASH_PAGES_PER_BANK\s+)\d+(\s*)",
-        r"(localparam\s+int\s+FLASH_PAGES_PER_BANK\s*=\s*)\d+(\s*;)",
-        r"({\s*name:\s*\"eflash\"," +
-        r"\s*clock_srcs:\s*{clk_i:\s*\"main\"}," +
-        r"\s*clock_group:\s*\"infra\"," +
-        r"\s*reset_connections:\s*{rst_ni:\s*\"lc\"}," +
-        r"\s*type:\s*\"eflash\"," +
-        r"\s*base_addr:\s*\"0x\w+\"," +
-        r"\s*swaccess:\s*\"ro\"," +
-        r"\s*size:\s*\"0x)\w+"
-    ]
-    replace = [
-        r"\g<1>32\g<2>",  # Change FLASH_PAGES_PER_BANK to 32
-        r"\g<1>32\g<2>",  # Change FLASH_PAGES_PER_BANK to 32
-        r"\g<1>10000"     # Change size to 0x10000
-    ]
-
-    # Change source files
-    for idx in range(len(files)):
-        with open(files[idx], 'r+') as file:
-            text = file.read()
-            text, num = re.subn(match[idx], replace[idx], text)
-            if num == 0:
-                print("ERROR: Cannot find regular expression " + match[idx] +
-                      " in " + files[idx] + ". "
-                      "Aborting")
-                sys.exit(1)
-            else:
-                print("Modifying " + files[idx])
-                file.seek(0)
-                file.write(text)
-                file.truncate()
+    # Modify hjson to change flash size
+    with open(top_hjson, "r") as hjson_file:
+        cfg = hjson.load(hjson_file,
+                         use_decimal=True)
+
+    # write out original version
+    # with open(top_path + "/original.hjson", "w") as hjson_file:
+    #     hjson.dump(cfg, hjson_file)
+
+    # update value
+    for mem in cfg["memory"]:
+        if mem['type'] == 'eflash':
+            mem['pages_per_bank'] = 32
+
+    # write back updated hjson
```

Comment:
Awesome, this will be much more stable than what I did :+1:

Reply:
Oh I see, that's not so straightforward. On one hand we want the script to be stable (pro hjson parse and rewrite) on the other hand it would be nice to have an easy readable diff (pro regexp search and replace). At the moment I don't see an easy solution to satisfy both requirements. But I value stability more.

What would be reasonable and helpful when debugging is a comment in the updated hjson at the very top saying that:
1. This file was rewritten by the reduce script and a list of the changed fields.
2. Due to the rewrite the diff gets messed up, but the rewritten original version is available too for the direct diff with the modified hjson (the original rewrite should be added to the .gitignore maybe).

WDYT?

Reply:
yep this sounds like a good idea, i've updated this.
I actually think we can take this approach to support different kinds of fpgas, especially since our real design will not fit on an fpga anyways.

So it makes sense to follow-up with a separate PR to more formally define how we want to change more than just the flash parameter. 

Let me know what you think.

Reply:
I have been working on a document to discuss pros and cons of different approaches for this. I will share that with you. Thanks to the templated flash and your work in this PR, we gained quite some flexibility.

I think we now should do some experiments to find out how much resources can actually be saved at what engineering effort/overhead. But let's continue this discussion in the document. It would be great to agree on a proposal before we share this in the whole project.